### PR TITLE
Feature: support nn-Meter in Proxyless NAS

### PR DESCRIPTION
**Pull request by JiahangXu**

### CODE REVIEW

File: nni/retiarii/converter/graph_gen.py [hunk 1]
```diff
@@ -1,6 +1,7 @@
 # Copyright (c) Microsoft Corporation.
 # Licensed under the MIT license.
 
+from os import replace
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:4]: why

File: nni/retiarii/converter/graph_gen.py [hunk 2]
```diff
@@ -249,6 +251,15 @@ def _generate_expr(tensor):
                     return f'({left} < {right})'
                 elif tensor.node().kind() == 'prim::If':
                     raise RuntimeError('Have not supported `if A and/or B`, please use two `if` statements instead.')
+                elif tensor.node().kind() == 'aten::abs':
```
Comment (ultmaster) [inline nni/retiarii/converter/graph_gen.py:254]: Suggest add tests for newly added operators and python name.

File: nni/retiarii/graph.py [hunk 3]
```diff
@@ -211,6 +211,28 @@ def get_node_by_name(self, node_name: str) -> 'Node':
...
+        for graph in self.graphs.values():
+            nodes = graph.get_nodes_by_python_name(python_name)
+            matched_nodes.extend(nodes)
+        # assert len(matched_nodes) <= 1
+        if matched_nodes:
+            return matched_nodes[0]
+        else:
+            return None
+
+    def get_layerchoice_nodes(self) -> List['Node']:
```
Comment (ultmaster) [inline nni/retiarii/graph.py:229]: Should not have `layerchoice` specific logic in graph. Suggest move it to graphgen.

File: nni/retiarii/graph.py [hunk 4]
```diff
@@ -366,6 +402,14 @@ def get_nodes_by_label(self, label: str) -> List['Node']:
 
     def get_nodes_by_name(self, name: str) -> List['Node']:
         return [node for node in self.hidden_nodes if node.name == name]
+    
+    def get_nodes_by_python_name(self, python_name: str) -> Optional['Node']:
+        return [node for node in self.nodes if node.python_name == python_name]
+
+    def get_layerchoice_nodes(self) -> List['Node']:
```
Comment (ultmaster) [inline nni/retiarii/graph.py:409]: Same here.

File: nni/retiarii/graph.py [hunk 5]
```diff
@@ -617,7 +668,15 @@ def _dump(self) -> Any:
             ret['operation']['cell_name'] = self.operation.cell_name
         if self.label is not None:
             ret['label'] = self.label
+        if self.python_name is not None:
+            ret['python_name'] = self.python_name
         return ret
+    
+    def set_python_name(self, python_name: str):
```
Comment (ultmaster) [inline nni/retiarii/graph.py:675]: I think `node.python_name = xxx` is enough. Does not necessarily need to call `set_python_name`

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 6]
```diff
@@ -209,7 +222,79 @@ def _train_one_epoch(self, epoch):
                 _logger.info("Epoch [%s/%s] Step [%s/%s]  %s", epoch + 1,
                              self.num_epochs, step + 1, len(self.train_loader), meters)
 
-    def _logits_and_loss(self, X, y):
+    def _form_latency_table(self, dummy_input):
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:225]: Can we put this into `LatencyPredictor`?

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 7]
```diff
@@ -209,7 +222,79 @@ def _train_one_epoch(self, epoch):
...
+        ''' return logits and loss for architecture parameter update '''
+        logits = self.model(X)
+        ce_loss = self.loss(logits, y)
+        if not self.block_latency_table:
+            return logits, ce_loss
+        # import time; since=time.time()
+        expected_latency = self._cal_expected_latency()
+        # print(f'_cal_expected_latency: {time.time() - since}')
+        
+        if self.reg_loss_type == 'mul#log':
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:278]: What does `mul#log` and `add#linear` mean?

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 8]
```diff
@@ -209,7 +222,79 @@ def _train_one_epoch(self, epoch):
...
+        
+        # form the latency of the stationary block in the latency table
+        temp_ir_model._root_graph_name = base_model_ir._root_graph_name
+        GraphConverterWithShape().flatten_without_layerchoice(temp_ir_model)
+        latency = self.latency_predictor.predict(temp_ir_model, model_type = 'nni-ir')
+        latency_table['stationary_block'] = {'root': latency}
+        
+        return latency_table
+
+    def _cal_expected_latency(self):
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:259]: This should be put into latency predictor as well.

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 9]
```diff
@@ -136,7 +139,10 @@ class ProxylessTrainer(BaseOneShotTrainer):
     def __init__(self, model, loss, metrics, optimizer,
                  num_epochs, dataset, warmup_epochs=0,
                  batch_size=64, workers=4, device=None, log_frequency=None,
-                 arc_learning_rate=1.0E-3):
+                 arc_learning_rate=1.0E-3, 
+                 grad_reg_loss_type=None, grad_reg_loss_params=None, 
+                 latency_predictor=None, dummy_input=(1, 3, 224, 224),
+                 ref_latency=65):
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:145]: What is ref latency?

File: nni/retiarii/oneshot/pytorch/proxyless.py (hunk 7 above)
Comment (JiahangXu) [reply nni/retiarii/oneshot/pytorch/proxyless.py:278]: The setting is following the origin setting in Proxyless NAS (https://github.com/mit-han-lab/proxylessnas/blob/6e7a96b7190963e404d1cf9b37a320501e62b0a0/search/nas_manager.py#L78). In their code, they provided two types of regularization loss by multiplication and addition, respectively. The reference latency is introduced here by the same reason, which represents an acceptable latency for users.

File: nni/retiarii/oneshot/pytorch/proxyless.py (hunk 9 above)
Comment (JiahangXu) [reply nni/retiarii/oneshot/pytorch/proxyless.py:145]: Refer to https://github.com/microsoft/nni/pull/4206#discussion_r716327695

File: nni/retiarii/graph.py (hunk 5 above)
Comment (JiahangXu) [reply nni/retiarii/graph.py:675]: Thanks for the comment. Will refine it.

File: nni/retiarii/converter/graph_gen.py (hunk 1 above)
Comment (JiahangXu) [reply nni/retiarii/converter/graph_gen.py:4]: Its a redundant import. My fault.

File: nni/retiarii/oneshot/pytorch/proxyless.py (hunk 6 above)
Comment (JiahangXu) [reply nni/retiarii/oneshot/pytorch/proxyless.py:225]: Do you mean define a class named `LatencyPredictor` and refer it to form latency table and get expected latency? Should it be a subclass of `ProxylessTrainer` or a separate class?

Comment (ultmaster) [reply nni/retiarii/oneshot/pytorch/proxyless.py:225]: I think you already have a class named LatencyPredictor?

I'm not sure. But putting latency-related logic in proxylessnas trainer looks very strange. If you have another trainer (let's say SPOS trainer or FBNet trainer or DARTS trainer or ENAS trainer) who has a similar need for latency calculation, do you need to write all these logic over again?

File: nni/retiarii/converter/graph_gen.py (hunk 2 above)
Comment (JiahangXu) [reply nni/retiarii/converter/graph_gen.py:253]: Got it. Will fix it soon.

File: examples/nas/oneshot/proxylessnas/ops.py [hunk 10]
```diff
@@ -270,7 +282,7 @@ def __init__(self, in_channels, out_channels,
             feature_dim = self.mid_channels
 
         if self.expand_ratio == 1:
-            self.inverted_bottleneck = None
+            self.inverted_bottleneck = nn.Sequential(OrderedDict([]))
```
Comment (ultmaster) [inline examples/nas/oneshot/proxylessnas/ops.py:285]: equivalent to `nn.Sequential()`

File: examples/nas/oneshot/proxylessnas/ops.py [hunk 11]
```diff
@@ -236,11 +249,10 @@ def bn_before_weight(self):
                 return True
             elif op == 'weight':
                 return False
-        raise ValueError('Invalid ops_order: %s' % self.ops_order)
+        raise ValueError(f'Invalid ops_order:{self.ops_order}')
```
Comment (ultmaster) [inline examples/nas/oneshot/proxylessnas/ops.py:252]: Sorry to be critical but you missed a "space".

File: nni/retiarii/graph.py [hunk 12]
```diff
@@ -212,6 +212,14 @@ def get_node_by_name(self, node_name: str) -> 'Node':
         else:
             return None
 
+    def get_cell_nodes(self) -> List['Node']:
+        matched_nodes = []
+        for graph in self.graphs.values():
+            nodes = [node for node in graph.nodes if isinstance(node.operation, Cell)]
+            matched_nodes.extend(nodes)
+        # assert len(matched_nodes) <= 1
```
Comment (ultmaster) [inline nni/retiarii/graph.py:220]: remove this

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 13]
```diff
@@ -136,7 +208,10 @@ class ProxylessTrainer(BaseOneShotTrainer):
     def __init__(self, model, loss, metrics, optimizer,
                  num_epochs, dataset, warmup_epochs=0,
                  batch_size=64, workers=4, device=None, log_frequency=None,
-                 arc_learning_rate=1.0E-3):
+                 arc_learning_rate=1.0E-3,
+                 grad_reg_loss_type=None, grad_reg_loss_params=None,
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:212]: Please add docstrings for newly-added parameters.

Comment (JiahangXu) [reply nni/retiarii/oneshot/pytorch/proxyless.py:212]: Have fixed other comments. The doc will commit tonight. Thanks a lot!

**Review by Lynazhang — APPROVED** [review]

**Review by liuzhe-lz — APPROVED** [review]

File: nni/retiarii/oneshot/pytorch/proxyless.py [hunk 14]
```diff
@@ -131,12 +203,31 @@ class ProxylessTrainer(BaseOneShotTrainer):
         Step count per logging.
     arc_learning_rate : float
         Learning rate of architecture parameters.
+    grad_reg_loss_type: string
+        Regularization type to add hardware related loss, allowed types include
+        - mul#log: regularized_loss = (torch.log(expected_latency) / math.log(self.ref_latency)) ** beta;
```
Comment (ultmaster) [inline nni/retiarii/oneshot/pytorch/proxyless.py:208]: Please wrap the code with ``` ``code`` ```

**Review by ultmaster — APPROVED** [review]
